deepinv/models/epll.py

```diff
@@ -0,0 +1,457 @@
+
+        :param torch.Tensor x: input data of shape batch_dimension x dimension
+        :param bool cov_regularization: whether using regularized covariance matrices
+        """
+        if cov_regularization:
+            cov_inv = self._cov_inv_reg
+            logdet_cov = self._logdet_cov_reg
+        else:
+            cov_inv = self._cov_inv
+            logdet_cov = self._logdet_cov
+        centered_x = x[None, :, :] - self.mu[:, None, :]
+        exponent = torch.sum(torch.bmm(centered_x, cov_inv) * centered_x, 2)
+        component_log_likelihoods = (
+            -0.5 * logdet_cov[:, None]
+            - 0.5 * exponent
+            - 0.5 * self.dimension * torch.log(torch.tensor(2 * torch.pi).to(x))
+        )
+        return component_log_likelihoods.T
+
+    def forward(self, x):
+        r"""
+        evaluate negative log likelihood function
+
+        :param torch.Tensor x: input data of shape batch_dimension x dimension
+        """
+        component_log_likelihoods = self.component_log_likelihoods(x)
+        component_log_likelihoods = component_log_likelihoods + torch.log(
+            self._weights[None, :]
+        )
+        log_likelihoods = torch.logsumexp(component_log_likelihoods, -1)
+        return -log_likelihoods
+
+    def classify(self, x, cov_regularization=False):
+        """
+        returns the index of the most likely component
+
+        :param torch.Tensor x: input data of shape batch_dimension x dimension
+        :param bool cov_regularization: whether using regularized covariance matrices
+        """
+        component_log_likelihoods = self.component_log_likelihoods(
+            x, cov_regularization=cov_regularization
+        )
+        component_log_likelihoods = component_log_likelihoods + torch.log(
+            self._weights[None, :]
+        )
+        val, ind = torch.max(component_log_likelihoods, 1)
+        return ind
+
+    def fit(
+        self,
+        dataloader,
+        max_iters=100,
+        stopping_criterion=None,
+        data_init=True,
+        cov_regularization=1e-5,
+        verbose=False,
+    ):
+        """
+        Batched Expectation Maximization algorithm for parameter estimation.
+
+
+        :param torch.utils.data.DataLoader dataloader: containing the data
+        :param int max_iters: maximum number of iterations
+        :param float stopping_criterion: stop when objective decrease is smaller than this number.
+            None for performing exactly max_iters iterations
+        :param bool data_init: True for initialize mu by the first data points, False for using current values as initialization
+        :param bool verbose: Output progress information in the console
+        """
+        if data_init:
+            first_data = next(iter(dataloader))[0][: self.n_components].to(self.mu)
+            if first_data.shape[0] == self.n_components:
+                # self.mu = first_data
+                self.mu.copy_(first_data)
```

Comment:
I've replaced line 387 with the following, is this ok? (it raised a bug in the demo file)